core/src/Streamly/Internal/FileSystem/Path.hs

```diff
@@ -0,0 +1,485 @@
+-- "File Name" and "Dir Name" to represent names. Also, file systems may put
+-- limits on names. Can have an IsName type class with members Name, (File
+-- Name), (Dir Name).
+
+-- | A type representing a file path.
+newtype File a = File a
+
+-- | A type representing a directory path.
+newtype Dir a = Dir a
+
+-- | A type representing absolute paths.
+newtype Abs a = Abs a
+
+-- | A type representing relative paths.
+newtype Rel a = Rel a
+
+------------------------------------------------------------------------------
+-- Conversions
+------------------------------------------------------------------------------
+
+-- | A member of 'IsPath' knows how to convert to and from the 'Path' type.
+class IsPath a where
+    -- | Like 'fromPath' but does not check the properties of 'Path'. Provides
+    -- performance and simplicity when we know that the properties of the path
+    -- are already verified, for example, when we get the path from the file
+    -- system or the OS APIs.
+    fromPathUnsafe :: Path -> a
+
+    -- XXX Use MonadThrow instead of Maybe?
+
+    -- | Convert a raw 'Path' to other forms of well-typed paths. It may fail
+    -- if the path does not satisfy the properties of the target type.
+    --
+    -- Path components may have limits.
+    -- Total path length may have a limit.
+    fromPath :: MonadThrow m => Path -> m a
+
+    -- | Convert a well-typed path to a raw 'Path'. Never fails.
+    toPath :: a -> Path
+
+instance IsPath Path where
+    fromPathUnsafe = id
+    fromPath = pure
+    toPath = id
+
+instance IsPath (File Path) where
+    fromPathUnsafe p = File p
+    fromPath p = pure (File p)
+    toPath (File p) = p
+
+instance IsPath (Dir Path) where
+    fromPathUnsafe p = Dir p
+    fromPath p = pure (Dir p)
+    toPath (Dir p) = p
+
+instance IsPath (Abs Path) where
+    fromPathUnsafe p = Abs p
+    fromPath p = pure (Abs p)
+    toPath (Abs p) = p
+
+instance IsPath (Rel Path) where
+    fromPathUnsafe p = Rel p
+    fromPath p = pure (Rel p)
+    toPath (Rel p) = p
+
+instance IsPath (Abs (File Path)) where
+    fromPathUnsafe p = Abs (File p)
+    fromPath p = pure (Abs (File p))
+    toPath (Abs (File p)) = p
+
+instance IsPath (Abs (Dir Path)) where
+    fromPathUnsafe p = Abs (Dir p)
+    fromPath p = pure (Abs (Dir p))
+    toPath (Abs (Dir p)) = p
+
+instance IsPath (Rel (File Path)) where
+    fromPathUnsafe p = Rel (File p)
+    fromPath p = pure (Rel (File p))
+    toPath (Rel (File p)) = p
+
+instance IsPath (Rel (Dir Path)) where
+    fromPathUnsafe p = Rel (Dir p)
+    fromPath p = pure (Rel (Dir p))
+    toPath (Rel (Dir p)) = p
+
+-- XXX Use rewrite rules to eliminate intermediate conversions for better
+-- efficiency.
+
+-- | Convert a path type to another path type. This operation may fail with a
+-- 'PathException' when converting a less restrictive path type to a more
+-- restrictive one.
+adaptPath :: (MonadThrow m, IsPath a, IsPath b) => a -> m b
+adaptPath p = fromPath $ toPath p
```

Comment:
With the current implementation this will never fail right?

Reply:
Without the checks, no. When we add the checks, it will.